src/eins-location.c

```diff
@@ -0,0 +1,197 @@
+  /* altitude is set to -1.7976931348623157e+308 if not determined */
+  gboolean altitude_known = (altitude > -1e308);
+
+  GVariant *location_info = g_variant_new ("(ddbdd)", latitude, longitude,
+                                           altitude_known, altitude, accuracy);
+
+  emtr_event_recorder_record_event (emtr_event_recorder_get_default (),
+                                    EVENT_USER_LOCATION, location_info);
+  /* Assumes ownership of location_info */
+
+  g_object_unref (location);
+}
+
+static void
+on_location_updated (GeoclueClient *client,
+                     gchar         *location_path_old,
+                     gchar         *location_path_new)
+{
+  geoclue_location_proxy_new_for_bus (G_BUS_TYPE_SYSTEM,
+                                      G_DBUS_PROXY_FLAGS_NONE,
+                                      "org.freedesktop.GeoClue2",
+                                      location_path_new,
+                                      NULL, /* cancellable */
+                                      (GAsyncReadyCallback) on_location_proxy_ready,
+                                      NULL /* data */);
+}
+
+static void
+on_start_ready (GeoclueClient *client,
+                GAsyncResult  *res)
+{
+  GError *error = NULL;
+
+  if (!geoclue_client_call_start_finish (client, res, &error))
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        g_critical ("Failed to start GeoClue2 client: %s", error->message);
+      g_error_free (error);
+    }
+}
+
+static void
+on_client_proxy_ready (GObject      *obj,
+                       GAsyncResult *res)
+{
+  GError *error = NULL;
+
+  GeoclueClient *client = geoclue_client_proxy_new_for_bus_finish (res, &error);
+  if (error != NULL)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        g_critical ("Failed to connect to GeoClue2 service: %s",
+                    error->message);
+      g_error_free (error);
+      return;
+    }
+
+  geoclue_client_set_desktop_id (client, DESKTOP_ID);
+  geoclue_client_set_distance_threshold (client, LOCATION_ACCURACY_CITY);
+  geoclue_client_set_requested_accuracy_level (client,
+                                               GCLUE_ACCURACY_LEVEL_CITY);
+
+  g_signal_connect (client, "location-updated",
+                    G_CALLBACK (on_location_updated), NULL);
+
+  geoclue_client_call_start (client, NULL, /* cancellable */
+                             (GAsyncReadyCallback) on_start_ready, NULL);
+}
+
+static void
+on_get_client_ready (GeoclueManager *manager,
+                     GAsyncResult   *res)
+{
+  gchar *client_path;
+  GError *error = NULL;
+
+  if (!geoclue_manager_call_get_client_finish (manager, &client_path, res,
+                                               &error))
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        g_critical ("Failed to connect to GeoClue2 service: %s",
```

Comment:
I would use a more specific message here mentioning we failed to get client from manager